File: packages/sdk/go/bugwatch_test.go
package bugwatch

import (
	"context"
	"errors"
	"sync"
	"testing"
)

// mockTransport implements Transport and stores captured events.
type mockTransport struct {
	mu     sync.Mutex
	events []*ErrorEvent
}

func (m *mockTransport) Send(event *ErrorEvent) error {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.events = append(m.events, event)
	return nil
}

func (m *mockTransport) getEvents() []*ErrorEvent {
	m.mu.Lock()
	defer m.mu.Unlock()
	out := make([]*ErrorEvent, len(m.events))
	copy(out, m.events)
	return out
}

func newTestClient(t *testing.T) (*BugwatchClient, *mockTransport) {
	t.Helper()
	mock := &mockTransport{}
	opts := &BugwatchOptions{
		ApiKey:      "test-key",
		Environment: "test",
		Release:     "1.0.0",
		ServerURL:   "http://localhost",
	}
	opts.applyDefaults()
	client := &BugwatchClient{
		options:   opts,
		transport: mock,
	}
	return client, mock
}

func TestInit(t *testing.T) {
	// Reset global client
	globalMu.Lock()
	globalClient = nil
	globalMu.Unlock()

	err := Init(&BugwatchOptions{
		ApiKey:    "test-key",
		ServerURL: "http://localhost",
	})
	if err != nil {
		t.Fatalf("Init failed: %v", err)
	}

	client := getClient()
	if client == nil {
		t.Fatal("Global client should not be nil after Init")
	}

	// Cleanup
	Close()
}

func TestInitWithoutApiKey(t *testing.T) {
	// Reset global client
	globalMu.Lock()
	globalClient = nil
	globalMu.Unlock()

	// Ensure env var is not set
	t.Setenv("BUGWATCH_API_KEY", "")

	err := Init(&BugwatchOptions{})
	if err == nil {
		t.Fatal("Init should fail without an API key")
	}
}

func TestCaptureException(t *testing.T) {
	client, mock := newTestClient(t)

	testErr := errors.New("something went wrong")
	eventID := client.CaptureException(testErr)

	if eventID == "" {
		t.Fatal("CaptureException should return a non-empty event ID")
	}

	events := mock.getEvents()
	if len(events) != 1 {
		t.Fatalf("Expected 1 event, got %d", len(events))
	}

	event := events[0]

	if event.ID != eventID {
		t.Errorf("Event ID mismatch: got %s, want %s", event.ID, eventID)
	}
	if event.Platform != "go" {
		t.Errorf("Platform should be 'go', got %s", event.Platform)
	}
	if event.Level != LevelError {
		t.Errorf("Level should be 'error', got %s", event.Level)
	}
	if event.Environment != "test" {
		t.Errorf("Environment should be 'test', got %s", event.Environment)
	}
	if event.Release != "1.0.0" {
		t.Errorf("Release should be '1.0.0', got %s", event.Release)
	}
	if event.Exception == nil {
		t.Fatal("Exception should not be nil")
	}
	if event.Exception.Value != "something went wrong" {
		t.Errorf("Exception value mismatch: got %s", event.Exception.Value)
	}
	if event.Fingerprint == "" {
		t.Error("Fingerprint should not be empty")
	}
	if event.Timestamp == "" {
		t.Error("Timestamp should not be empty")
	}
}

func TestCaptureExceptionNil(t *testing.T) {
	client, mock := newTestClient(t)

	eventID := client.CaptureException(nil)

	if eventID != "" {
		t.Error("CaptureException(nil) should return empty string")
	}
	if len(mock.getEvents()) != 0 {
		t.Error("No event should be sent for nil error")
	}
}

func TestCaptureMessage(t *testing.T) {
	client, mock := newTestClient(t)

	eventID := client.CaptureMessage("test message", LevelWarning)

	if eventID == "" {
		t.Fatal("CaptureMessage should return a non-empty event ID")
	}

	events := mock.getEvents()
	if len(events) != 1 {
		t.Fatalf("Expected 1 event, got %d", len(events))
	}

	event := events[0]

	if event.Level != LevelWarning {
		t.Errorf("Level should be 'warning', got %s", event.Level)
	}
	if event.Exception == nil {
		t.Fatal("Exception should not be nil")
	}
	if event.Exception.Type != "Message" {
		t.Errorf("Exception type should be 'Message', got %s", event.Exception.Type)
	}
	if event.Exception.Value != "test message" {
		t.Errorf("Exception value should be 'test message', got %s", event.Exception.Value)
	}
}

func TestCaptureExceptionWithContext(t *testing.T) {
	client, mock := newTestClient(t)

	ctx := context.Background()
	ctx = WithUser(ctx, &UserContext{
		ID:    "user-123",
		Email: "test@example.com",
	})
	ctx = WithBreadcrumb(ctx, Breadcrumb{
		Message:  "navigated",
		Category: "navigation",
	})
	ctx = WithRequestContext(ctx, &RequestContext{
		URL:    "/api/test",
		Method: "GET",
	})

	testErr := errors.New("context error")
	eventID := client.CaptureExceptionWithContext(ctx, testErr)

	if eventID == "" {
		t.Fatal("Should return non-empty event ID")
	}

	events := mock.getEvents()
	event := events[0]

	if event.User == nil {
		t.Fatal("User context should be set")
	}
	if event.User.ID != "user-123" {
		t.Errorf("User ID should be 'user-123', got %s", event.User.ID)
	}
	if event.User.Email != "test@example.com" {
		t.Errorf("User email mismatch")
	}
	if len(event.Breadcrumbs) != 1 {
		t.Fatalf("Expected 1 breadcrumb, got %d", len(event.Breadcrumbs))
	}
	if event.Breadcrumbs[0].Message != "navigated" {
		t.Errorf("Breadcrumb message mismatch")
	}
	if event.Request == nil {
		t.Fatal("Request context should be set")
	}
	if event.Request.URL != "/api/test" {
		t.Errorf("Request URL should be '/api/test', got %s", event.Request.URL)
	}
}

func TestBeforeSendDrop(t *testing.T) {
	mock := &mockTransport{}
	opts := &BugwatchOptions{
		ApiKey:    "test-key",
		ServerURL: "http://localhost",
		BeforeSend: func(event *ErrorEvent) *ErrorEvent {
			return nil // drop all events
		},
	}
	opts.applyDefaults()
	client := &BugwatchClient{
		options:   opts,
		transport: mock,
	}

	eventID := client.CaptureException(errors.New("dropped"))

	if eventID != "" {
		t.Error("Should return empty string when BeforeSend drops event")
	}
	if len(mock.getEvents()) != 0 {
		t.Error("No event should be sent when BeforeSend returns nil")
	}
}

func TestBeforeSendModify(t *testing.T) {
	mock := &mockTransport{}
	opts := &BugwatchOptions{
		ApiKey:    "test-key",
		ServerURL: "http://localhost",
		BeforeSend: func(event *ErrorEvent) *ErrorEvent {
			event.Tags = map[string]string{"modified": "true"}
			return event
		},
	}
	opts.applyDefaults()
	client := &BugwatchClient{
		options:   opts,
		transport: mock,
	}

	client.CaptureException(errors.New("modified"))

	events := mock.getEvents()
	if len(events) != 1 {
		t.Fatalf("Expected 1 event, got %d", len(events))
	}
	if events[0].Tags["modified"] != "true" {
		t.Error("BeforeSend should have added the 'modified' tag")
	}
}

File: packages/sdk/java/core/src/test/java/dev/bugwatch/BugwatchTest.java
package dev.bugwatch;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class BugwatchTest {

    private final List<ErrorEvent> capturedEvents = new ArrayList<>();

    private final Transport mockTransport = new Transport() {
        @Override
        public boolean send(ErrorEvent event) {
            capturedEvents.add(event);
            return true;
        }
    };

    @BeforeEach
    void setUp() {
        capturedEvents.clear();
        BugwatchOptions options = BugwatchOptions.builder()
                .apiKey("test-api-key")
                .environment("test")
                .release("1.0.0")
                .build();
        BugwatchClient client = new BugwatchClient(options, mockTransport);
        Bugwatch.init(client);
    }

    @AfterEach
    void tearDown() {
        Bugwatch.close();
    }

    @Test
    void testInit() {
        assertNotNull(Bugwatch.getClient());
    }

    @Test
    void testCaptureException() {
        RuntimeException exception = new RuntimeException("Test error");
        String eventId = Bugwatch.captureException(exception);

        assertNotNull(eventId);
        assertEquals(1, capturedEvents.size());

        ErrorEvent event = capturedEvents.get(0);
        assertEquals(eventId, event.getId());
        assertEquals("error", event.getLevel());
        assertEquals("java", event.getPlatform());
        assertEquals("test", event.getEnvironment());
        assertEquals("1.0.0", event.getRelease());
        assertNotNull(event.getException());

        ExceptionInfo ex = event.getException();
        assertEquals("java.lang.RuntimeException", ex.type());
        assertEquals("Test error", ex.value());
        assertNotNull(ex.stacktrace());
        assertFalse(ex.stacktrace().isEmpty());
    }

    @Test
    void testCaptureExceptionWithCause() {
        Exception cause = new IllegalArgumentException("root cause");
        RuntimeException exception = new RuntimeException("wrapper", cause);
        String eventId = Bugwatch.captureException(exception);

        assertNotNull(eventId);
        assertEquals(1, capturedEvents.size());

        ErrorEvent event = capturedEvents.get(0);
        assertNotNull(event.getException());
        assertEquals("java.lang.RuntimeException", event.getException().type());
    }

    @Test
    void testCaptureMessage() {
        String eventId = Bugwatch.captureMessage("Something happened", Level.WARNING);

        assertNotNull(eventId);
        assertEquals(1, capturedEvents.size());

        ErrorEvent event = capturedEvents.get(0);
        assertEquals(eventId, event.getId());
        assertEquals("warning", event.getLevel());
        assertNotNull(event.getException());
        assertEquals("Something happened", event.getException().value());
    }

    @Test
    void testCaptureWithoutInit() {
        Bugwatch.close();
        String eventId = Bugwatch.captureException(new RuntimeException("test"));
        assertNull(eventId);
    }

    @Test
    void testBeforeSendDrop() {
        Bugwatch.close();
        capturedEvents.clear();

        BugwatchOptions options = BugwatchOptions.builder()
                .apiKey("test-api-key")
                .environment("test")
                .beforeSend(event -> null) // Drop all events
                .build();
        BugwatchClient client = new BugwatchClient(options, mockTransport);
        Bugwatch.init(client);

        String eventId = Bugwatch.captureException(new RuntimeException("dropped"));
        assertNull(eventId);
        assertTrue(capturedEvents.isEmpty());
    }

    @Test
    void testBeforeSendModify() {
        Bugwatch.close();
        capturedEvents.clear();

        BugwatchOptions options = BugwatchOptions.builder()
                .apiKey("test-api-key")
                .environment("test")
                .beforeSend(event -> {
                    event.addTag("custom", "tag");
                    return event;
                })
                .build();
        BugwatchClient client = new BugwatchClient(options, mockTransport);
        Bugwatch.init(client);

        Bugwatch.captureException(new RuntimeException("modified"));
        assertEquals(1, capturedEvents.size());
        assertEquals("tag", capturedEvents.get(0).getTags().get("custom"));
    }

    @Test
    void testFingerprint() {
        Bugwatch.captureException(new RuntimeException("Test error"));
        assertNotNull(capturedEvents.get(0).getFingerprint());
        assertFalse(capturedEvents.get(0).getFingerprint().isBlank());
    }

    @Test
    void testScopeManager() {
        BugwatchClient client = Bugwatch.getClient();
        assertNotNull(client);

        client.getScopeManager().setUser(new UserContext("u1", "test@example.com", "tester", "127.0.0.1"));
        Bugwatch.captureException(new RuntimeException("scoped error"));

        assertEquals(1, capturedEvents.size());
        assertNotNull(capturedEvents.get(0).getUser());
        assertEquals("u1", capturedEvents.get(0).getUser().id());
    }
}
